internal/istio/client.go

```diff
@@ -0,0 +1,112 @@
+package istio
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	istioclientapiv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	istioclient "istio.io/client-go/pkg/clientset/versioned"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+)
+
+type Client struct {
+	istioclient.Interface
+	eventRecorder record.EventRecorder
+	logger        logr.Logger
+}
+
+func NewIstioClient(cfg *rest.Config, recorder record.EventRecorder,
+	logger logr.Logger,
+) (*Client, error) {
+	cs, err := istioclient.NewForConfig(cfg)
+	if err != nil {
+		return nil, errors.Join(ErrFailedToCreateIstioClient, err)
+	}
+	return &Client{
+		Interface:     cs,
+		eventRecorder: recorder,
+		logger:        logger,
+	}, nil
+}
+
+func (c *Client) GetVirtualService(ctx context.Context, name, namespace string) (*istioclientapiv1beta1.VirtualService, error) {
+	virtualService, err := c.NetworkingV1beta1().
+		VirtualServices(namespace).
+		Get(ctx, name, apimetav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Join(ErrFailedToGetVirtualService, err)
+	}
+	return virtualService, nil
+}
+
+func (c *Client) ListVirtualServices(ctx context.Context, namespace string) (*istioclientapiv1beta1.VirtualServiceList, error) {
+	virtualServiceList, err := c.NetworkingV1beta1().
+		VirtualServices(namespace).
+		List(ctx, apimetav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Join(ErrFailedToListVirtualServices, err)
+	}
+	return virtualServiceList, nil
+}
+
+func (c *Client) CreateVirtualService(ctx context.Context, virtualService *istioclientapiv1beta1.VirtualService) error {
+	_, err := c.NetworkingV1beta1().
+		VirtualServices(virtualService.GetNamespace()).
+		Create(ctx, virtualService, apimetav1.CreateOptions{})
+	if err != nil {
+		return errors.Join(ErrFailedToCreateVirtualService, err)
+	}
+	return nil
+}
+
+func (c *Client) UpdateVirtualService(ctx context.Context, virtualService,
+	virtualServiceRemote *istioclientapiv1beta1.VirtualService,
+) error {
+	virtualServiceRemote.Name = virtualService.Name
+	virtualServiceRemote.Namespace = virtualService.Namespace
+	virtualService.Spec.DeepCopyInto(&virtualServiceRemote.Spec)
+
+	_, err := c.NetworkingV1beta1().
+		VirtualServices(virtualServiceRemote.Namespace).
+		Update(ctx, virtualServiceRemote, apimetav1.UpdateOptions{})
+	if err != nil {
+		return errors.Join(ErrFailedToUpdateVirtualService, err)
+	}
+	return nil
+}
+
+func (c *Client) DeleteVirtualService(ctx context.Context, name, namespace string,
+) error {
+	err := c.NetworkingV1beta1().
+		VirtualServices(namespace).
+		Delete(ctx, name, apimetav1.DeleteOptions{})
+	if err != nil {
+		return errors.Join(ErrFailedToDeleteVirtualService, err)
+	}
+	return nil
+}
+
+func (c *Client) ListGatewaysByLabelSelector(ctx context.Context, labelSelector *apimetav1.LabelSelector) (*istioclientapiv1beta1.GatewayList,
```

Comment:
This was previously `GatewayLookup` which received the watcher and returned the list of gateway items. I opted for directly using the LabelSelector argument now since we were only getting this exact thing from the watcher CR, and return the `GatewayList` instead for consistency with the other list functions.